anesthetic/utils.py

```diff
@@ -182,3 +183,131 @@ def iso_probability_contours(pdf, contours=[0.68, 0.95]):
                 c[j] = c[j] - 1e-5
 
     return c
+
+
+def iso_probability_contours_from_samples(pdf, contours=[0.68, 0.95],
+                                          weights=None):
+    """Compute the iso-probability contour values."""
+    if weights is None:
+        weights = numpy.ones_like(pdf)
+    contours = [1-p for p in reversed(contours)]
+    i = numpy.argsort(pdf)
+    m = numpy.cumsum(weights[i])
+    m /= m[-1]
+    interp = interp1d([0]+list(m), [0]+list(pdf[i]))
+    c = list(interp(contours))+[max(pdf)]
+
+    # Correct level sets
+    for i in range(1, len(c)):
+        if c[i-1] == c[i]:
+            for j in range(i):
+                c[j] = c[j] - 1e-5
+
+    return c
+
+
+def triangular_sample_compression_2d(x, y, w=None, n=1000):
+    """Histogram a 2D set of weighted samples via triangulation.
+
+    This defines bins via a triangulation of the subsamples, sums weights
+    within triangles, and computes weighted centroids of triangles.
+
+    Parameters
+    ----------
+    x, y: array-like
+        x and y coordinates of samples for compressing
+
+    w: array-like, optional
+        weights of samples
+
+    n: int, optional
+        number of samples returned. Default 1000
+
+    Returns
+    -------
+    x, y, w, array-like
+        Compressed samples and weights
+
+    """
+    if w is None:
+        w = numpy.ones_like(x)
+
+    # Select samples for triangulation
+    if sum(w != 0) < n:
+        i = numpy.arange(len(w))
+    else:
+        i = numpy.random.choice(len(w), size=n, replace=False, p=w/w.sum())
```

Comment:
Should `replace` really be `False` here? 
How does that affect the weighting?

Reply:
Excellent question. It all depends on what you do with the weights after. For concreteness, consider this code:

```python
x = numpy.array([1,2,3])
w = numpy.array([1e10,1,1])
print(numpy.random.choice(x, size=2, replace=False, p = w/w.sum()))
```

You are right that `replace=False` would definitely be wrong if you were using it to 'de-weight' a set of samples (as in the `MCMCSamples.compress`) function). However, if you keep the weights, then this acts as an alternative form of compression . You can actually think of the strategy performed in `MCMCSamples.compress` as doing this, but with the optimal `size` chosen by the channel capacity.

In actual fact, since we are only using this to bin the weights into ~O(1000) samples, it doesn't even matter that i is chosen as a compression strategy. For example, if you wanted to ensure better coverage in the tails, the `p` in the above line need not be proportional to the weights. Points chosen in the tails would aggregate to have very low weight, but that may be what you want in practice if you wanted to go beyond two-sigma contours.